File: core/src/rustyice/game/lights/components/LightComponent.java
package rustyice.game.lights.components;

import box2dLight.Light;
import box2dLight.RayHandler;
import com.badlogic.gdx.graphics.Color;
import rustyice.editor.annotations.ComponentProperty;
import rustyice.game.GameObject;
import rustyice.game.physics.FillterFlags;

public abstract class LightComponent {
    static final float LIGHT_RES = 1f;
    private static final Color DEFAULT_COLOR = new Color(.75f, .75f, .75f, .75f);

    private transient boolean initialized = false;
    private transient Light light;
    private transient GameObject parent;

    private Color color;
    private boolean isStatic;
    private boolean isXRay;
    private float distance;
    private float direction;
    private float x, y;

    public LightComponent(){
        color = DEFAULT_COLOR;
        isStatic = false;
        isXRay = false;
        distance = 10;
    }

    @ComponentProperty(title = "Color")
    public Color getColor() {
        return color;
    }

    @ComponentProperty(title = "Color")
    public void setColor(Color color) {
        this.color = color;
        if(initialized){
            light.setColor(color);
        }
    }

    public void setPosition(float x, float y){
        this.x = x;
        this.y = y;
        if(initialized){
            light.setPosition(x, y);
        }
    }

    public float getX(){
        return x;
    }

    public float getY(){
        return y;
    }

    public void setDirection(float direction){
        this.direction = direction;
        if(initialized){
            light.setDirection(direction);
        }
    }


    public float getDirection(){
        return direction;
    }

    @ComponentProperty(title = "Static")
    public void setStaticLight(boolean isStatic) {
        this.isStatic = isStatic;
        if(initialized){
            light.setStaticLight(isStatic);
        }
    }

    @ComponentProperty(title = "Static")
    public boolean isStaticLight() {
        return isStatic;
    }

    @ComponentProperty(title = "XRay")
    public void setXRay(boolean xRay){
        isXRay = xRay;
        if(initialized){
            light.setXray(isXRay);
        }
    }

    @ComponentProperty(title = "XRay")
    public boolean isXRay(){
        return isXRay;
    }

    @ComponentProperty(title = "Distance")
    public float getDistance() {
        return distance;
    }

    @ComponentProperty(title = "Distance")
    public void setDistance(float distance) {
        this.distance = distance;
        rebuild();
    }

    protected abstract Light buildLight(RayHandler rayHandler);

    public boolean isInitialized(){
        return initialized;
    }

    public void init(GameObject parent){
        if(!initialized){
            initialized = true;
            this.parent = parent;

            light = buildLight(parent.getSection().getRayHandler());

            light.setContactFilter(FillterFlags.LIGHT, (short)0, FillterFlags.OPAQUE);
            light.setColor(color);
            light.setDistance(distance);
            light.setPosition(x, y);
            light.setStaticLight(isStatic);
            light.setXray(isXRay);
            light.setDirection(direction);

            //light.setSoft(false);
        }
    }

    public void rebuild(){
        if(isInitialized()){
            store();
            init(parent);
        }
    }

    public void store(){
        if(initialized){
            initialized = false;

            light.remove();
            light = null;
        }
    }
}

File: core/src/rustyice/game/lights/LightComponent.kt
package rustyice.game.lights

import box2dLight.Light
import box2dLight.RayHandler
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import rustyice.editor.annotations.ComponentAccess
import rustyice.editor.annotations.ComponentProperty
import rustyice.game.GameLifecycle
import rustyice.game.GameObject
import rustyice.game.MissingParentReference
import rustyice.graphics.Camera
import rustyice.physics.LIGHT
import rustyice.physics.OPAQUE

abstract class LightComponent: GameLifecycle() {
    val LIGHT_RES = 1f
    val DEFAULT_COLOR = Color(.75f, .75f, .75f, .75f)

    @Transient private var light: Light? = null
    @Transient var parent: GameObject? = null

    @ComponentProperty
    var color = DEFAULT_COLOR
        set(value) {
            field = value
            light?.color = value
        }

    @ComponentProperty
    var static = false
        set(value){
            field = value
            light?.isStaticLight = value
        }

    @ComponentProperty
    var xRay = false
        set(value){
            field = value
            light?.isXray = value
        }

    @ComponentProperty
    var distance = 10f
        set(value){
            field = value
            reInit()
        }
    var direction = 0f
        set(value) {
            field = value
            light?.direction = value
        }
    var x = 0f
        set(value){
            field = value
            light?.setPosition(value, y)
        }
    var y = 0f
        set(value){
            field = value
            light?.setPosition(x, value)
        }

    fun setPosition(x: Float, y: Float){
        this.x = x
        this.y = y
        light?.setPosition(x, y)
    }


    protected abstract fun buildLight(rayHandler: RayHandler): Light

    override fun init(){
        super.init()

        val rayHandler = parent?.game?.lightingHandler
            ?: throw MissingParentReference("Game")

        val light = buildLight(rayHandler)

        light.setContactFilter(LIGHT.toShort(), 0, OPAQUE.toShort())
        light.color = color
        light.distance = distance
        light.setPosition(x, y)
        light.isStaticLight = static
        light.isXray = xRay
        light.direction = direction

        this.light = light
    }

    override fun update(delta: Float) {}

    override fun render(batch: Batch, camera: Camera, renderFlags: Int) {}

    override fun store(){
        super.store()

        light?.remove()
        light = null
    }

    interface LightContainer{
        @ComponentAccess
        val lightComponent: LightComponent
    }
}
